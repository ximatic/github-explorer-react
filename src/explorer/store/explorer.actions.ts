export enum ExplorerAction {
  Reset = 'explorer/reset',
  //ResetToken = "explorer/resetToken",
  VerifyToken = 'explorer/verifyToken',
  VerifyTokenSuccess = 'explorer/verifyTokenSuccess',
  VerifyTokenError = 'explorer/verifyTokenError',
  //Token = 'explorer/token',
  LoadRepositories = 'explorer/loadRepositories',
  LoadRepositoriesSuccess = 'explorer/loadRepositoriesSuccess',
  LoadRepositoriesError = 'explorer/loadRepositoriesError',
  LoadRepository = 'explorer/loadRepository',
  LoadRepositorySuccess = 'explorer/loadRepositorySuccess',
  LoadRepositoryError = 'explorer/loadRepositoryError',
  PageInfo = 'explorer/pageInfo',
}
