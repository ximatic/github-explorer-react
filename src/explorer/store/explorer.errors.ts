export enum ExplorerError {
  InvalidToken = 'EXPLORER_ERROR_INVALID_TOKEN',
}
