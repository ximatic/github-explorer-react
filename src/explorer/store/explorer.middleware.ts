import { Middleware } from '@reduxjs/toolkit';

import { ApolloQueryResult, gql } from '@apollo/client';

import { apolloClient } from '../../ApolloProvider';

import { GraphqlPageInfo, GraphqlRepositoriesData, GraphqlRepositoriesNode } from '../models/explorer-graphql.model';
import { ExplorerPageInfo, Repository } from '../models/explorer.model';

import { ExplorerAction } from './explorer.actions';
import { loadRepositoriesError, loadRepositoriesSuccess, pageInfo, verifyTokenError, verifyTokenSuccess } from './explorer.slice';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const explorerMiddleware: Middleware = (storeAPI: any) => (next: any) => (action: any) => {
  const result = next(action);

  const storageKey = 'gh-explorer-token';

  // token

  function storeToken(token: string): void {
    localStorage.setItem(storageKey, JSON.stringify(token));
  }

  function resetToken(): void {
    localStorage.removeItem(storageKey);
  }

  // page info

  function createPageInfo(itemsPerPage: number, pageInfo: GraphqlPageInfo): ExplorerPageInfo {
    return {
      itemsPerPage,
      cursorStart: pageInfo.startCursor,
      cursorEnd: pageInfo.endCursor,
      hasNextPage: pageInfo.hasNextPage,
      hasPreviousPage: pageInfo.hasPreviousPage,
    };
  }

  switch (action.type) {
    case ExplorerAction.Reset:
      resetToken();
      break;
    case ExplorerAction.VerifyToken:
      try {
        const token = action.payload.token;
        fetch('https://api.github.com/octocat', {
          headers: {
            Authorization: `Bearer ${token}`,
          },
        })
          .then(async (data: Response) => {
            if (!data.ok) {
              storeAPI.dispatch(verifyTokenError());
              return;
            }
            if (action.payload.storeToken) {
              storeToken(token);
            }
            storeAPI.dispatch(verifyTokenSuccess({ token }));
          })
          .catch(() => {
            storeAPI.dispatch(verifyTokenError());
          });
      } catch {
        storeAPI.dispatch(verifyTokenError());
      }
      break;
    case ExplorerAction.LoadRepositories:
      try {
        const pagination = action.payload.pagination;
        const token = storeAPI.getState().explorer.token;

        apolloClient
          .query({
            query: gql`
                query {
                  search(query: "is:public sort:stars-desc", type: REPOSITORY, ${pagination.itemsKey}: ${pagination.itemsValue}, ${pagination.cursorKey}: ${pagination.cursorValue}) {
                    nodes {
                      ...repositoryFragment
                    }
                    pageInfo {
                      endCursor
                      startCursor
                      hasNextPage
                      hasPreviousPage
                    }
                  }
                }
                fragment repositoryFragment on Repository {
                  name
                  createdAt
                  owner {
                    login
                  }
                  description
                  stargazerCount
                }
              `,
            context: {
              headers: {
                Authorization: `Bearer ${token}`,
              },
            },
          })
          .then((result: ApolloQueryResult<GraphqlRepositoriesData>) => {
            const repositories = result.data.search.nodes.map((node: GraphqlRepositoriesNode) => {
              return {
                name: node.name,
                owner: node.owner?.login || 'N/A',
                stars: node.stargazerCount,
                createdAt: node.createdAt,
                description: node.description,
              } as Repository;
            });

            storeAPI.dispatch(loadRepositoriesSuccess({ repositories }));
            storeAPI.dispatch(pageInfo({ pageInfo: createPageInfo(pagination.itemsValue, result.data.search.pageInfo) }));
          });
      } catch {
        storeAPI.dispatch(loadRepositoriesError());
      }
      break;
  }

  return result;
};
