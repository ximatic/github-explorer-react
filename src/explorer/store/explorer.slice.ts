import { createSlice, ReducerCreators, type PayloadAction } from '@reduxjs/toolkit';

import { ExplorerEvent, ExplorerEventName, ExplorerEventType, ExplorerState } from './explorer.state';
import { ExplorerPageInfo, ExplorerPagination, Repository } from '../models/explorer.model';

const initialState: ExplorerState = {
  token: null,
  repositories: null,
  repository: null,
  pageInfo: null,
  error: null,
  event: null,
};

export const explorerSlice = createSlice({
  name: 'explorer',
  initialState: initialState,
  reducers: (create: ReducerCreators<ExplorerState>) => ({
    // reset
    reset: create.reducer(() => initialState),
    // verify token
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    verifyToken: create.reducer((state: ExplorerState, action: PayloadAction<{ token: string; storeToken: boolean }>) => {
      state.token = initialState.token;
      state.repositories = initialState.repositories;
      state.repository = initialState.repository;
      state.pageInfo = initialState.pageInfo;
      state.event = {
        name: ExplorerEventName.VerifyToken,
        type: ExplorerEventType.Processing,
      };
    }),
    verifyTokenSuccess: create.reducer((state: ExplorerState, action: PayloadAction<{ token: string }>) => {
      state.token = action.payload.token;
      state.repositories = initialState.repositories;
      state.repository = initialState.repository;
      state.pageInfo = initialState.pageInfo;
      state.event = {
        name: ExplorerEventName.VerifyToken,
        type: ExplorerEventType.Success,
      };
    }),
    verifyTokenError: create.reducer((state: ExplorerState) => {
      state.token = initialState.token;
      state.repositories = initialState.repositories;
      state.repository = initialState.repository;
      state.pageInfo = initialState.pageInfo;
      state.event = {
        name: ExplorerEventName.VerifyToken,
        type: ExplorerEventType.Error,
      };
    }),
    // load repositories
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    loadRepositories: create.reducer((state: ExplorerState, action: PayloadAction<{ pagination: ExplorerPagination }>) => {
      state.repository = initialState.repository;
      state.pageInfo = initialState.pageInfo;
      state.event = {
        name: ExplorerEventName.LoadRepositories,
        type: ExplorerEventType.Processing,
      };
    }),
    loadRepositoriesSuccess: create.reducer((state: ExplorerState, action: PayloadAction<{ repositories: Repository[] }>) => {
      state.repositories = action.payload.repositories;
      state.repository = initialState.repository;
      state.event = {
        name: ExplorerEventName.LoadRepositories,
        type: ExplorerEventType.Success,
      };
    }),
    loadRepositoriesError: create.reducer((state: ExplorerState) => {
      state.repositories = initialState.repositories;
      state.repository = initialState.repository;
      state.pageInfo = initialState.pageInfo;
      state.event = {
        name: ExplorerEventName.LoadRepositories,
        type: ExplorerEventType.Error,
      };
    }),
    // load repository
    loadRepository: create.reducer(
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      (state: ExplorerState, action: PayloadAction<{ owner: string; name: string; pagination: ExplorerPagination }>) => {
        state.pageInfo = initialState.pageInfo;
        state.event = {
          name: ExplorerEventName.LoadRepository,
          type: ExplorerEventType.Processing,
        };
      },
    ),
    loadRepositorySuccess: create.reducer((state: ExplorerState, action: PayloadAction<{ repository: Repository }>) => {
      state.repository = action.payload.repository;
      state.event = {
        name: ExplorerEventName.LoadRepository,
        type: ExplorerEventType.Success,
      };
    }),
    loadRepositoryError: create.reducer((state: ExplorerState) => {
      state.repository = initialState.repository;
      state.pageInfo = initialState.pageInfo;
      state.event = {
        name: ExplorerEventName.LoadRepository,
        type: ExplorerEventType.Error,
      };
    }),
    // pagination
    pageInfo: create.reducer((state: ExplorerState, action: PayloadAction<{ pageInfo: ExplorerPageInfo }>) => {
      state.pageInfo = action.payload.pageInfo;
    }),
  }),
  selectors: {
    selectExplorerEvent: (state: ExplorerState): ExplorerEvent | null => state.event,
    selectToken: (state: ExplorerState): string | null => state.token,
    selectRepositories: (state: ExplorerState): Repository[] | null => state.repositories,
    selectRepository: (state: ExplorerState): Repository | null => state.repository,
    selectPageInfo: (state: ExplorerState): ExplorerPageInfo | null => state.pageInfo,
  },
});

export const {
  reset,
  verifyToken,
  verifyTokenSuccess,
  verifyTokenError,
  loadRepositories,
  loadRepositoriesSuccess,
  loadRepositoriesError,
  loadRepository,
  loadRepositorySuccess,
  loadRepositoryError,
  pageInfo,
} = explorerSlice.actions;

export const { selectExplorerEvent, selectToken, selectRepositories, selectRepository, selectPageInfo } = explorerSlice.selectors;
