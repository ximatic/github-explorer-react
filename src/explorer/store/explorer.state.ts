import { Repository, ExplorerPageInfo } from '../models/explorer.model';
import { ExplorerError } from './explorer.errors';

export enum ExplorerEventName {
  VerifyToken = 'verify-token',
  LoadRepositories = 'load-repositories',
  LoadRepository = 'load-repository',
}

export enum ExplorerEventType {
  Processing = 'processing',
  Success = 'success',
  Error = 'error',
}

export interface ExplorerEvent {
  name: ExplorerEventName;
  type: ExplorerEventType;
  message?: string;
}

export interface ExplorerState {
  token: string | null;
  repositories: Repository[] | null;
  repository: Repository | null;
  pageInfo: ExplorerPageInfo | null;
  error: ExplorerError | null;
  event: ExplorerEvent | null;
}
