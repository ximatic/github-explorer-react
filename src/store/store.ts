import { configureStore } from '@reduxjs/toolkit';

import { explorerSlice } from '../explorer/store/explorer.slice';
import { explorerMiddleware } from '../explorer/store/explorer.middleware';

// export const store = configureStore({
//   reducer: {
//     explorer: explorerSlice.reducer,
//   },
//   middleware: (getDefaultMiddleware) => {
//     return getDefaultMiddleware().concat(explorerMiddleware);
//   },
// });

// export type AppStore = typeof store;
// export type AppDispatch = typeof store.dispatch;
// export type RootState = ReturnType<typeof store.getState>;

export const makeStore = () => {
  return configureStore({
    reducer: {
      explorer: explorerSlice.reducer,
    },
    middleware: (getDefaultMiddleware) => {
      return getDefaultMiddleware().concat(explorerMiddleware);
    },
  });
};

// Infer the type of makeStore
export type AppStore = ReturnType<typeof makeStore>;
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<AppStore['getState']>;
export type AppDispatch = AppStore['dispatch'];
