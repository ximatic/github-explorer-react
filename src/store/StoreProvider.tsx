'use client';

import { useRef } from 'react';
import { Provider } from 'react-redux';

import { AppStore, makeStore } from './store';

interface StoreProviderProps {
  readonly children: React.ReactNode;
}

export default function StoreProvider({ children }: StoreProviderProps) {
  const storeRef = useRef<AppStore | null>(null);
  if (!storeRef.current) {
    // Create the store instance the first time this renders
    storeRef.current = makeStore();
  }

  return <Provider store={storeRef.current}>{children}</Provider>;
}
