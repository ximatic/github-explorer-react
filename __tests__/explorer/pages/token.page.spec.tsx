import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import * as router from 'react-router-dom';

import { useAppDispatch, useAppSelector } from '../../../src/store/hooks';

import { ExplorerEventName, ExplorerEventType } from '../../../src/explorer/store/explorer.state';
import { verifyToken } from '../../../src/explorer/store/explorer.slice';

import TokenPage from '../../../src/explorer/pages/token.page';

jest.mock('../../../src/store/hooks');
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

const mockUseAppDispatch = useAppDispatch as unknown as jest.Mock;
const mockUseAppSelector = useAppSelector as unknown as jest.Mock;
const mockUseNavigate = router.useNavigate as jest.Mock;

describe('TokenPage', () => {
  let dispatchMock: jest.Mock;
  let navigateMock: jest.Mock;

  const renderComponent = () => render(<TokenPage />);

  beforeEach(() => {
    dispatchMock = jest.fn();
    navigateMock = jest.fn();

    mockUseAppDispatch.mockReturnValue(dispatchMock);
    mockUseNavigate.mockReturnValue(navigateMock);
    mockUseAppSelector.mockReturnValue(undefined);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  // snapshot tests

  test('matches snapshot', () => {
    const { asFragment } = renderComponent();

    expect(asFragment()).toMatchSnapshot();
  });

  // unit tests

  it('renders welcome and form UI', () => {
    renderComponent();

    expect(screen.getByText(/GitHub Explorer/i)).toBeInTheDocument();
    expect(screen.getByText(/Your API Token/i)).toBeInTheDocument();
  });

  it('disables submit button when token is empty', () => {
    renderComponent();

    const button = screen.getByRole('button', { name: /start browsing/i });

    expect(button).toBeDisabled();
  });

  it('enables submit and dispatches verifyToken when token is entered', () => {
    renderComponent();

    const input = screen.getByLabelText(/token/i);
    fireEvent.change(input, { target: { value: 'abc123' } });
    const button = screen.getByRole('button', { name: /start browsing/i });
    fireEvent.click(button);

    expect(button).not.toBeDisabled();
    expect(dispatchMock).toHaveBeenCalledWith(verifyToken({ token: 'abc123', storeToken: false }));
  });

  it('shows error message when invalid token event is received', () => {
    mockUseAppSelector.mockReturnValue({
      name: ExplorerEventName.VerifyToken,
      type: ExplorerEventType.Error,
    });

    renderComponent();

    const input = screen.getByLabelText(/token/i);
    fireEvent.change(input, { target: { value: 'badtoken' } });
    fireEvent.input(input);

    expect(screen.getByText(/Provided API Token is invalid/i)).toBeInTheDocument();
  });

  it('navigates to /repositories on successful token verification', () => {
    mockUseAppSelector.mockReturnValue({
      name: ExplorerEventName.VerifyToken,
      type: ExplorerEventType.Success,
    });

    renderComponent();

    expect(navigateMock).toHaveBeenCalledWith('/repositories');
  });
});
