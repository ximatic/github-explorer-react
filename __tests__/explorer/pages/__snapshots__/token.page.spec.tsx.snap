// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenPage matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="dashboard"
  >
    <div
      class="p-panel--border mb-4 block p-panel p-component"
      data-pc-name="panel"
      data-pc-section="root"
      id="pr_id_1"
    >
      <div
        aria-hidden="false"
        aria-labelledby="pr_id_1_header"
        class="p-toggleable-content"
        data-pc-section="toggleablecontent"
        id="pr_id_1_content"
        role="region"
      >
        <div
          class="p-panel-content"
          data-pc-section="content"
        >
          <div
            class="grid place-items-center"
          >
            <p>
              <span
                class="pi pi-github !text-5xl"
              />
            </p>
            <p
              class="pt-2"
            >
              Welcome to 
              <strong>
                GitHub Explorer
              </strong>
            </p>
            <p
              class="pt-2 text-center"
            >
              Please follow the intructions from panel below to start browsing GitHub public repositories.
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="mb-4 block p-panel p-component"
      data-pc-name="panel"
      data-pc-section="root"
      id="pr_id_2"
    >
      <div
        class="p-panel-header"
        data-pc-section="header"
      >
        <span
          class="p-panel-title"
          data-pc-section="title"
          id="pr_id_2_header"
        >
          GitHub Authentication
        </span>
        <div
          class="p-panel-icons"
          data-pc-section="icons"
        />
      </div>
      <div
        aria-hidden="false"
        aria-labelledby="pr_id_2_header"
        class="p-toggleable-content"
        data-pc-section="toggleablecontent"
        id="pr_id_2_content"
        role="region"
      >
        <div
          class="p-panel-content"
          data-pc-section="content"
        >
          <form>
            <div
              class="mb-3"
            >
              Your API Token*
            </div>
            <div>
              <span
                class="p-float-label"
                data-pc-name="floatlabel"
                data-pc-section="root"
              >
                <input
                  class="w-full p-inputtext p-component"
                  data-pc-name="inputtext"
                  data-pc-section="root"
                  id="token"
                  value=""
                />
                <label
                  for="token"
                >
                  Token
                </label>
              </span>
            </div>
            <div
              class="mb-3 mt-3 flex items-center"
            >
              <div
                class="p-checkbox p-component"
                data-p-disabled="false"
                data-p-highlight="false"
                data-pc-name="checkbox"
                data-pc-section="root"
              >
                <input
                  aria-invalid="false"
                  class="p-checkbox-input"
                  data-pc-section="input"
                  type="checkbox"
                />
                <div
                  class="p-checkbox-box"
                  data-p-disabled="false"
                  data-p-highlight="false"
                  data-pc-section="box"
                />
              </div>
              <div
                class="ml-2"
              >
                Store token locally**
              </div>
            </div>
          </form>
          <div
            class="mt-4"
          >
            <button
              class="block w-full p-button p-component p-disabled"
              data-pc-name="button"
              data-pc-section="root"
              disabled=""
            >
              <span>
                Start browsing
              </span>
            </button>
          </div>
          <div
            class="mt-4"
          >
            <small
              class="block"
            >
              * By default, API Token is stored only temporarly on your machine as long as you are on this page.
            </small>
            <small
              class="block"
            >
              ** API Token can be stored locally in the browser (using 
              <code>
                localStorage
              </code>
              ) to simplify usage.
            </small>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
